Update training.py and sklearn.py for evals_result

JohanManders opened the pull request:
Made changes to training.py to make sure all eval_metric information get passed to evals_result. Previous version lost and mislabeled data in evals_result when using more than one eval_metric.

Structure of eval_metric is now:
```
eval_metric[evals][eval_metric] = list of metrics
```
Example for XGBoost train:
```python
import xgboost as xgb

dtrain = xgb.DMatrix('../data/agaricus.txt.train', silent=True)
dtest = xgb.DMatrix('../data/agaricus.txt.test', silent=True)

param = [('max_depth', 2), ('objective', 'binary:logistic'), ('bst:eta', 0.01), ('eval_metric', 'logloss'), ('eval_metric', 'error')]

watchlist  = [(dtest,'eval'), (dtrain,'train')]
num_round = 2
evals_result = {}
bst = xgb.train(param, dtrain, num_round, watchlist, evals_result=evals_result)

print('Access logloss metric directly from eval:')
print(evals_result['eval']['logloss'])

print('')

print('Access metrics through a loop:')
for e_name, e_mtrs in evals_result.items():
    print('- {}'.format(e_name))
    for e_mtr_name, e_mtr_vals in e_mtrs.items():
        print('   - {}'.format(e_mtr_name))
        print('      - {}'.format(e_mtr_vals))

print('')

print('Access complete dict:')
print(evals_result)
```
Prints:
```
Access logloss metric directly from eval:
['0.480385', '0.357756']

Access metrics through a loop:
- train
   - logloss
      - ['0.48253', '0.35953']
- eval
   - logloss
      - ['0.480385', '0.357756']

Access complete dict:
{'train': {'logloss': ['0.48253', '0.35953']}, 'eval': {'logloss': ['0.480385', '0.357756']}}
```

Also made changes to sklearn.py to make sure the structure of evals_result is the same as training.py.

Example for Sklearn implementation:
```python
import xgboost as xgb
import numpy as np
from sklearn.datasets import make_hastie_10_2

X, y = make_hastie_10_2(n_samples=2000, random_state=42)

# map labels from {-1, 1} to {0, 1}
labels, y = np.unique(y, return_inverse=True)

X_train, X_test = X[:1600], X[1600:]
y_train, y_test = y[:1600], y[1600:]

param_dist = {'objective':'binary:logistic', 'n_estimators':2}

clf = xgb.XGBModel(**param_dist)

# You can also use: 
# clf = xgb.XGBClassifier(**param_dist)
 
clf.fit(X_train, y_train,
        eval_set=[(X_train, y_train), (X_test, y_test)], 
        eval_metric='logloss',
        verbose=True)
 
evals_result = clf.evals_result()
 
print('Access logloss metric directly from validation_0:')
print(evals_result['validation_0']['logloss'])
 
print('')
 
print('Access metrics through a loop:')
for e_name, e_mtrs in evals_result.items():
    print('- {}'.format(e_name))
    for e_mtr_name, e_mtr_vals in e_mtrs.items():
        print('   - {}'.format(e_mtr_name))
        print('      - {}'.format(e_mtr_vals))
 
print('')
 
print('Access complete dict:')
print(evals_result)
```
Prints:
```
Access logloss metric directly from validation_0:
['0.682260', '0.672904']

Access metrics through a loop:
- validation_0
   - logloss
      - ['0.682260', '0.672904']
- validation_1
   - logloss
      - ['0.68558', '0.67917']

Access complete dict:
{'validation_0': {'logloss': ['0.682260', '0.672904']}, 'validation_1': {'logloss': ['0.68558', '0.67917']}}
```

tqchen commented:
Thanks!
-  Please update the docstring of eval_result to give a bit more detail of the return values and what is supposed to passed in. 
- Besides that put your name to https://github.com/dmlc/xgboost/blob/master/CONTRIBUTORS.md 

I will merge it after the update.

JohanManders commented:
-	I changed sklearn.py to make sure the evals_result is consistent when using training.py and sklearn.py
-	I added a small update to the docstring in training.py 
-	I changed the way you call the evals_result in sklearn.py (using a function instead of directly accasing the variable) and added an example to the docstings.
-	I added two demos to the guide-python folder for more clarity and edited the readme.md in this folder.
-	I added me to the contributers.md
-	All checks have passed

I don’t know if you want more info in the training.py docstring, or that the demos are fine. Please let me know.

tqchen commented:
Thanks! Sorry for the delayed action. This is now merged
